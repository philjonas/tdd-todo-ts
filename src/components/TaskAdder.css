.task-adder{
    display: flex;
    justify-content: center;
}