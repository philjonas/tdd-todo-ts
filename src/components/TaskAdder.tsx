import React, { ChangeEvent, useState } from "react";
import "./TaskAdder.css";

export interface TaskAdderProps {
  addTask?: (val: string) => void;
  initialValue?: string;
}

export const TaskAdder = ({
  addTask,
  initialValue = "",
}: TaskAdderProps): JSX.Element => {
  const [value, setValue] = useState(initialValue);

  const handleChange = (ev: ChangeEvent<HTMLInputElement>) => {
    ev.preventDefault();
    setValue(ev.currentTarget.value);
  };

  const clearText = () => setValue("");

  const handleAdd = () => {
    addTask?.(value);
    clearText();
  };

  return (
    <div className="task-adder">
      <div>
        Task:
        <input
          value={value}
          data-testid="task-adder-text"
          type="text"
          placeholder="write your task here"
          onChange={handleChange}
        />
      </div>
      <div>
        <button data-testid="task-adder-add" onClick={handleAdd}>
          Add Task
        </button>
        <button data-testid="task-adder-clear" onClick={clearText}>
          Clear Text
        </button>
      </div>
    </div>
  );
};
