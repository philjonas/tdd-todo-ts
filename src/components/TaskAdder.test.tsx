import { render, fireEvent } from "@testing-library/react";
import { TaskAdder, TaskAdderProps } from "./TaskAdder";

const setup = (props?: TaskAdderProps) => {
  const utils = render(<TaskAdder {...props} />);
  const input = utils.getByTestId("task-adder-text");
  const add = utils.getByTestId("task-adder-add");
  const clear = utils.getByTestId("task-adder-clear");
  return {
    input,
    add,
    clear,
    ...utils,
  };
};

describe("TaskAdder", () => {
  it("renders correctly", () => {
    const wrapper = <TaskAdder />;
    expect(wrapper).toMatchSnapshot();
  });

  it("has a placeholder", () => {
    const { input } = setup();
    expect(input.getAttribute("placeholder")).toBe("write your task here");
  });

  it("should update the value", () => {
    const { input } = setup();
    fireEvent.change(input, { target: { value: "123" } });
    expect(input.getAttribute("value")).toBe("123");
  });

  it("should clear the value", () => {
    const { input, clear } = setup();
    fireEvent.change(input, { target: { value: "123" } });
    expect(input.getAttribute("value")).toBe("123");
    fireEvent.click(clear);
    expect(input.getAttribute("value")).toBe("");
  });

  it("should fire add event when add button clicked", () => {
    const props: TaskAdderProps = { addTask: jest.fn() };
    const { add, input } = setup(props);
    fireEvent.change(input, { target: { value: "123" } });
    expect(input.getAttribute("value")).toBe("123");
    fireEvent.click(add);
    expect(props.addTask).toBeCalledTimes(1);
    expect(input.getAttribute("value")).toBe("");
  });
});
