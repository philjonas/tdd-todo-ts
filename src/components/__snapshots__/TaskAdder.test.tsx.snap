// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskAdder renders correctly 1`] = `<TaskAdder />`;
